packages/adapter-d1/src/conversion.ts

```diff
@@ -28,43 +30,70 @@ export function getColumnTypes(columnNames: string[], rows: Object[]): ColumnTyp
   return columnTypes as ColumnType[]
 }
 
-// JavaScript	D1
-// null	NULL
-// Number	REAL
-// Number 1	INTEGER
-// String	TEXT
-// Boolean 2	INTEGER
-// ArrayBuffer	BLOB
-// undefined	Not supported. Queries with undefined values will return a D1_TYPE_ERROR
-//
-// 1 D1 supports 64-bit signed INTEGER values internally, however BigInts
-// are not currently supported in the API yet. JavaScript integers are safe up to Number.MAX_SAFE_INTEGER
-//
-// 2 Booleans will be cast to an INTEGER type where 1 is TRUE and 0 is FALSE.
-
+/**
+ * Default mapping between JS and D1 types.
+ * | JavaScript       | D1           |
+ * | :--------------: | :---------:  |
+ * | null             | NULL         |
+ * | Number           | REAL         |
+ * | Number¹          | INTEGER      |
+ * | null             | TEXT         |
+ * | Boolean²         | INTEGER      |
+ * | ArrayBuffer      | BLOB         |
+ *
+ * ¹ - D1 supports 64-bit signed INTEGER values internally, however BigInts are not currently supported in the API yet. JavaScript integers are safe up to Number.MAX_SAFE_INTEGER.
+ *
+ * ² - Booleans will be cast to an INTEGER type where 1 is TRUE and 0 is FALSE.
+ */
 function inferColumnType(value: NonNullable<Value>): ColumnType {
   switch (typeof value) {
     case 'string':
-      return ColumnTypeEnum.Text
-    // case 'bigint':
-    //   return ColumnTypeEnum.Int64
-    // case 'boolean':
-    //   return ColumnTypeEnum.Boolean
+      return inferStringType(value)
     case 'number':
-      // Hack - TODO change this when we have type metadata
-      if (Number.isInteger(value) && Math.abs(value) < Number.MAX_SAFE_INTEGER) {
-        return ColumnTypeEnum.Int32
-      } else {
-        return ColumnTypeEnum.UnknownNumber
-      }
+      return inferNumberType(value)
     case 'object':
       return inferObjectType(value)
     default:
       throw new UnexpectedTypeError(value)
   }
 }
 
-function inferObjectType(value: {}): ColumnType {
+// See https://stackoverflow.com/a/3143231/1345244
+const isoDateRegex = new RegExp(
+  /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
+)
+function isISODate(str) {
+  return isoDateRegex.test(str)
```

Comment:
I'd prefer `new Date(str) !== 'Invalid Date'`. It lets in a little bit more than only ISO dates, but no regex.

Reply:
But we have no idea what's in the string, so it would result in weird things like for example
```js
new Date('Hello world 5')
-> Date Tue May 01 2001 00:00:00 GMT+0200 (Central European Summer Time)
```